Feature/dss12 sc 161126 plugin add jinja templating to write body

liamlynch-data opened the pull request:
~SHOULD NOT BE MERGED until all 12.6 relevant work on this plugin is done~
~(we want to keep master clear for urgent fixes - we can merge later work into this PR)~
(OK now as v0.20 has been tagged)

Adds templating and send as HTML functionality - also has some refactoring and other minor changes.

## Refactoring
- Factored out an email sending client and STMP config (should simplify things when we add the use of use channels to send emails)
- Factored out attachment dataset handling code from the recipe itself

## Send as HTML 
- Added a format choice param
- Added an HTML body param
- These are separate fields but only one of them is shown,  depending on the format choice  

## Templating
- See comment in SC card for how data is named and accessed from template
- The possibility of collision with the name `attachments` is handled by giving the contacts ds priority - if a a collision is detected, this is logged with a warn level allowing the user to debug the case of such a collision hading the attachments data

- Added JINJA templating - passed in column values from contacts DB so can be used when message is rendered

in  attachments_template_dict()
- made an HTML table of the first 50 rows of each attachment ds available too
- made a dict of the first 50 rows of actual data also available - used a std pandas format for this   

- Added default values for HTML and plain text body params with std CSS and some brief hints about JINJA 
- Added description with more instructions on how reference the different types of data

# Other
- changed the order of params and some labels/descriptions as agreed with TPM  

# Perf/optimisation
- Only the first 50 rows are used for display for  attachment DSs
- I experimented with  partial replacing the template once the attachment data is calculated, so the templating engine would not have to do so much work each time. But this did not seem to be possible (or at least not easy) with JINJA.

liamlynch-data committed:
increment version to 1.0.0

liamlynch-data committed:
refactoring - added python-lib and factoring things out

liamlynch-data committed:
added .idea to .gitignore

liamlynch-data committed:
add jinja dependency [sc-161126]

liamlynch-data committed:
add some templating functionality (WIP) [sc-161126]

liamlynch-data committed:
finished templating functionality, reorg of params in ui, add desc info and default body values [sc-161126]

liamlynch-data committed:
default value should be utf-8 [sc-161126]

liamlynch-data committed:
removed html table bits from plain text body description as not useful there [sc-161126]

liamlynch-data committed:
change default of use_html_body to false as otherwise exist configs are misintepreted [sc-161126]

liamlynch-data committed:
tweak to changelog [sc-161126]

liamlynch-data committed:
turn format choice into a select for better legacy recipe handling [sc-161126]

liamlynch-data committed:
handle and log collisions with variable called attachments [sc-161126]

liamlynch-data committed:
improved some variable naming and deleted some dead code [sc-161126]

alexbourret reviewed (approved):
All good. Tested using mailtrap.

liamlynch-data committed:
some refactor and prep for channel sending [sc-161131]

liamlynch-data committed:
remove body_encoding param [sc-161131]

liamlynch-data committed:
added channel client to send to channels [sc-161131]

liamlynch-data committed:
small tidy ups [sc-161131]

liamlynch-data committed:
use key of null for SMTP direct option to support legacy [sc-161131]

liamlynch-data committed:
adjust to match small API changes [sc-161131]

liamlynch-data committed:
moving smtp auth section to second section so it is obvious when it appears

liamlynch-data committed:
added dynamic channel select list [sc-161131]

liamlynch-data committed:
get sender field to disappear when needed, changes to match API, fixes to format handling [sc-161131]

liamlynch-data committed:
added some logging for the messaging clients

liamlynch-data committed:
fixed some html/css issues and some validation problems

liamlynch-data committed:
added made a 'subject' variable available for the template

liamlynch-data committed:
improved error handling

liamlynch-data committed:
added info to changelog, improved error handling [sc-161131]

liamlynch-data committed:
handle legacy recipes and DSS < 12.6 by conditionally defaulting to direct SMTP [sc-161131]

liamlynch-data committed:
removing 'Channel:' from select names

liamlynch-data committed:
changing plugin to match API changes

liamlynch-data committed:
move smtp settings, clean up use of channel API

liamlynch-data committed:
remove exposing subject as a templating var - was a misunderstanding

liamlynch-data committed:
add templating within sender field

liamlynch-data committed:
Updating the recipe description

liamlynch-data committed:
git ignoring something

liamlynch-data committed:
fix namedtype issues, were broken for python 3.6

liamlynch-data committed:
updated jinja and python requirements and changlelog to match

liamlynch-data committed:
simplify description as it was a bit big

liamlynch-data committed:
changing some file names for best practice

liamlynch-data committed:
add validation for recipient column (there and matches one in the contacts ds) [sc-161131]

liamlynch-data committed:
fix to DSSSelectorChoices (has no impact here but just in case)

liamlynch-data committed:
fix bug regarding body in column always sending in html [sc-170779]

liamlynch-data committed:
Merge pull request #9 from dataiku/feature/dss12-sc-161131-plugin-leverage-the-api-email-channel

Adding capability to send to DSS channels [sc-161131]

MayeulRousselet commented on code-env/python/spec/requirements.txt:
```diff
@@ -1 +1,2 @@
 XlsxWriter>=0.7.7
+Jinja2>=3.0.3
```
Better to pin dependencies version to avoid conflict on dependencies updates

liamlynch-data replied:
done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/e66f281b2c53eb71f24a221f14b3a39ea65f4a27)

MayeulRousselet replied:
Thanks, have you tried the dependencies compat with all the supported python versions?

liamlynch-data replied:
We will retest on supported py versions when PR comments addressed and other PR merged in. 
Will note that when done - I'll leaving this comment open as a reminder to do it

liamlynch-data replied:
Tested plugin and code env installation (and functionality, briefly) with all supported versions of python: namely 3.6, 3.7, 3.8, 3.9, 3.10, 3.11

MayeulRousselet commented on python-lib/dku_email_client.py:
```diff
@@ -0,0 +1,112 @@
+from email.mime.multipart import MIMEMultipart
+
+
+def SmtpConfig():
```
Why not using a class here? I find that using a NamedTuple is confusing when creating it:
 ```return SmtpConfig()(smtp_host, smtp_port, smtp_use_tls, smtp_use_auth, smtp_user, smtp_pass)```
One can forget the empty parenthesis...
Simpler to have:
```return SmtpConfig(smtp_host, smtp_port, smtp_use_tls, smtp_use_auth, smtp_user, smtp_pass)```

liamlynch-data replied:
done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/d94564d35ffc214777d231bc91872c8d123a4760

MayeulRousselet commented on python-lib/dku_email_client.py:
```diff
@@ -0,0 +1,112 @@
+    return namedtuple("SmtpConfig", "smtp_host, smtp_port, smtp_use_tls, smtp_use_auth, smtp_user, smtp_pass")
+
+
+def AttachmentFile():
```
Same comment as above

liamlynch-data replied:
done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/d94564d35ffc214777d231bc91872c8d123a4760

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -1,13 +1,85 @@
+    Sends the message or throws an exception
+    """
+
+    recipient = contacts_row[recipient_column]
```
`recipient_column` should be a method parameter, not a global variable (defined at https://github.com/dataiku/dss-plugin-sendmail/blob/0f7673b296f5228559d51fe8220a8826afb34c14/custom-recipes/send-mails-from-contacts-dataset/recipe.py#L96).
The main reasons are:
- for clarity, each method state should be explicit
- in case of a refactor (for instancve moving the method in a different module), it will break

liamlynch-data replied:
Done - the use of all global variables in functions was removed in https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/999759e91f8bebfd727e45cb21449a9d275a3bdd

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -1,13 +1,85 @@
+    recipient = contacts_row[recipient_column]
+    templating_value_dict = dict(contacts_row)
+
+    if use_subject_value:
```
Should a method parameter (see previous comment)

liamlynch-data replied:
Done - the use of all global variables in functions was removed in https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/999759e91f8bebfd727e45cb21449a9d275a3bdd

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -1,13 +1,85 @@
+    else:
+        email_subject = contacts_row.get(subject_column, "")
+
+    if use_body_value:
```
should be a method parameter (see previous comment)

liamlynch-data replied:
Done - the use of all global variables in functions was removed in https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/999759e91f8bebfd727e45cb21449a9d275a3bdd

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -1,13 +1,85 @@
+        email_text = contacts_row.get(body_column, "")
+
+    # Note - if the channel has a sender configured, the sender value will be ignored by the email client here
+    sender = sender_value if use_sender_value else contacts_row.get(sender_column, "")
```
sender_value should be a parameter

liamlynch-data replied:
Done - the use of all global variables in functions was removed in https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/999759e91f8bebfd727e45cb21449a9d275a3bdd

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -1,13 +1,85 @@
+    return channel_id is not None and channel_id.endswith(SENDER_SUFFIX)
+
+
+def send_email_for_contact(mail_client, contacts_row, message_template, subject_line_template):
```
This method should be reworked.
First, it uses global variables declared outside of its scope.
Secondly, for testability, it will preferable to have a method `build_email` that construct a email object, and then call a `mail_cllient.send(email)`. Thus it would be very simple to unit test the method `build_email`

liamlynch-data replied:
Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/999759e91f8bebfd727e45cb21449a9d275a3bdd) and [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/016bc7b31bdf078ef242d76a367b11f3dba8cbfb)

Then gave the functions better names in https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/60eb12fe40e3f2008f36d64fe00624bd2a583d03

Two utility functions were pulled out, `build_email_subject` and `build_email_message_text`

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -1,13 +1,85 @@
+    mail_client.send_email(sender, recipient, email_subject, email_text, attachment_files)
+
 
 logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s %(message)s')
```
Logging logic should go on top of the file, just after the import

liamlynch-data replied:
done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/a21dd452a024158265115637d2338a47103402a8)

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -31,88 +103,84 @@
 people_columns = [p['name'] for p in people.read_schema()]
-for arg in ['sender', 'subject', 'body']:
+for arg in ['subject', 'body']:
     if not globals()["use_" + arg + "_value"] and globals()[arg + "_column"] not in people_columns:
```
Overkill (not your code) but ok

liamlynch-data replied:
I agree but I'll leave it - low ROI to address this

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -31,88 +103,84 @@
-        mime_parts.append(app)
+body_template = None
+if use_body_value:
+    if use_html_body_value:
```
Per definition, `use_html_body_value = use_body_value and (body_format == 'html')`, so the first test is redondant. 
better to have:
```
if use_body_value:
    if body_format == 'html':
       ....
    else
       ... 
```

liamlynch-data replied:
https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/c49da5ae3fd7e6a33e51e25bcc02847f0c2654f1

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -122,7 +190,7 @@ def send_email(contact):
             try:
-                send_email(contact)
+                send_email_for_contact(email_client, contact, body_template, subject_template)
                 d = dict(contact)
```
in `send_email_for_contact`, contact is already transformed to a dict (https://github.com/dataiku/dss-plugin-sendmail/blob/0f7673b296f5228559d51fe8220a8826afb34c14/custom-recipes/send-mails-from-contacts-dataset/recipe.py#L47). so it will be transformed 2 times. Better to transform it before, to pass it as a dict in the method

liamlynch-data replied:
I changed that [here ](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/a21dd452a024158265115637d2338a47103402a8)

But note I need to clone the dict anyway as `templating_value_dict` can have an attachments key and we don't want to write it to the output dataset

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -31,88 +103,84 @@
+# Validation part 1 - Check some kind of value/column exists for body, subject, sender and recipient
 
-if not body_column and not body_value:
+has_body_column = body_column and not use_body_value
+has_plain_body_value = use_body_value and body_value and not use_html_body_value
+has_html_body_value = use_body_value and use_html_body_value and html_body_value
+if not (has_body_column or has_plain_body_value or has_html_body_value):
```
Very complex to follow, can't we find something easier to read?

liamlynch-data replied:
I see what you mean. When I was writing it I thought it was clear as there are 3 UI states we need to check for validity and this made that very explicit. But now I realise that it would be clearer to do in if statements as follows - this would reflect the choices present in the UI and is easier to follow I think - do you agree?

```
# Check a body is provided for the options selected
is_body_present = False
if use_body_value:
    if use_html_body_value:
        is_body_present = bool(html_body_value)
    else:
        is_body_present = bool(body_value)
else:
    is_body_present = bool(body_column)

if not is_body_present:
    raise AttributeError("No body column nor body value specified")
```

liamlynch-data replied:
Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/e4fe6e3751e9e11b47a4d603404955e57311514e)

(Definitely clearer this way)

MayeulRousselet commented on resource/dynamic_channels.py:
```diff
@@ -0,0 +1,29 @@
+        # Add an entry for direct SMTP
+        if len(channels) > 0:
+            # If there is a choice of channels, giving direct SMTP a key of "direct_smtp" means it is there but not as default
+            choices.append("Manually define SMTP", 'direct_smtp')
```
⚠️ : `direct_smtp` could be a id that is given to a channel by an admin.
To avoid id collision, please use `__DKU__DIRECT_SMTP__`

liamlynch-data replied:
True! Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/380a7c0d9cc8de378e2e09664890e9828501ce6f)

MayeulRousselet commented on resource/dynamic_channels.py:
```diff
@@ -0,0 +1,29 @@
+        dss_client = dataiku.api_client()
+        channels = []
+        # To work for < 12.6 DSS - first check API to list channels exists, only call it if it does
+        if callable(getattr(dss_client, "list_messaging_channels", None)):
```
Nitpick: This should be in a dedicated method call `support_messaging_channels` for clarity

liamlynch-data replied:
This change is already lined up in the follow up PR [here](https://github.com/dataiku/dss-plugin-sendmail/pull/10/files#diff-44f227161ef56923c43a70ade7493cd01df31e6e43d33fba892abba4021da9b4)

liamlynch-data replied:
Which was merged in [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/1e727f41eb737bbc3ccb347f060123fec1065cc0#diff-44f227161ef56923c43a70ade7493cd01df31e6e43d33fba892abba4021da9b4)

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.json:
```diff
@@ -48,21 +98,23 @@
             "columnRole" : "contacts",
             "description" : "Sender of the email (from a column)",
-            "visibilityCondition" : "!model.use_sender_value"
+            "visibilityCondition" : "!model.use_sender_value && !model.mail_channel.endsWith('__|*_S')"
```
⚠️ : The "sender exists" suffix is very hard to gasp as it appears as a RegExp whereas it is not a RegExp, it is only a fixed suffix.
Better to use a more explicit suffix, such as `__WITH_DEFINED_SENDER__`

liamlynch-data replied:
Ok - I was trying to minimise the risk of collision, but you're right, it's confusing! I'll change it

liamlynch-data replied:
Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/1896899111bbb8be3aefdf39a05d8b0357a38082)

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.json:
```diff
@@ -48,21 +98,23 @@
             "type": "STRING",
             "description" : "Sender of the  email (provide value inline)",
-            "visibilityCondition" : "model.use_sender_value"
+            "visibilityCondition" : "model.use_sender_value && !model.mail_channel.endsWith('__|*_S')"
```
Same comment as above

liamlynch-data replied:
Done above

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.json:
```diff
@@ -48,21 +98,23 @@
+            "label" : " ",
+            "description" : "Use custom value",
+            "type": "BOOLEAN",
+            "visibilityCondition" : "!model.mail_channel.endsWith('__|*_S')"
```
Same comment as above

liamlynch-data replied:
Done above

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -1,13 +1,85 @@
+    smtp_use_auth = recipe_config.get('smtp_use_auth', False)
+    smtp_user = recipe_config.get('smtp_user', None)
+    smtp_pass = recipe_config.get('smtp_pass', None)
+    return SmtpConfig()(smtp_host, smtp_port, smtp_use_tls, smtp_use_auth, smtp_user, smtp_pass)
```
Better to not use NamedTuple, see comment on python-lib/dku_email_client.py:12

liamlynch-data replied:
done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/d94564d35ffc214777d231bc91872c8d123a4760

liamlynch-data committed:
adding optioin for excel conditional formating [sc-161134]

liamlynch-data committed:
making the apply conditional formating only appear for DSS 12.6 onwards [sc-161134]

liamlynch-data committed:
changed name of dynamic channels as no longer just about channels [sc-161134]

MayeulRousselet commented on custom-recipes/send-mails-from-contacts-dataset/recipe.py:
```diff
@@ -122,7 +190,7 @@ def send_email(contact):
         for contact in people.iter_rows():
             logging.info("Sending to %s" % contact)
```
Not your code but false: 
`contact` is the entire row, so logging `Sending to %s` will log the entire row as a json in the logs. 
It was not a big deal before the updates, but as we can have a lot of columns, body etc... in the contact dataset I think we should only log the recipient (as the log message should do)

![Capture d’écran 2024-02-20 à 6 24 48 PM](https://github.com/dataiku/dss-plugin-sendmail/assets/54799490/301a99c7-0f85-4536-b8b3-46cd29061415)

liamlynch-data replied:
Ok - I will change it to just log the recipient if present, and the whole row if absent (the email will fail in that case - I will not change how that is handled)
```
            recipient = contact[recipient_column]
            if recipient:
                logging.info("Sending to %s" % contact[recipient_column])
            else:
                logging.info("No recipient for row - emailing will fail - row data: %s" % contact)
```                
                
In the logs this looks like this:

```
2024-03-01 11:43:18,111 INFO Sending to c@test.com
2024-03-01 11:43:18,821 INFO Sending to d@test.com
2024-03-01 11:43:19,595 INFO No recipient for row - emailing will fail - row data: {'Name': 'Mr Missing', 'Email': '', 'Sender': '', 'Email body': '', 'Balance': None, 'Status': '', 'Special_offer': False, 'attachments_____': ''}
2024-03-01 11:43:19,701 ERROR Send failed
Traceback (most recent call last):
  File "<string>", line 200, in <module>
  File "<string>", line 83, in send_email_for_contact
  File "/Users/liamlynch/devenvs_auto/devenv_8122/dss-home/plugins/dev/sendmail/python-lib/dku_email_client.py", line 59, in send_email
    self.send_email_impl(sender, recipient, email_subject, email_body_to_send, attachment_files)
  File "/Users/liamlynch/devenvs_auto/devenv_8122/dss-home/plugins/dev/sendmail/python-lib/dku_email_client.py", line 130, in send_email_impl
    self.smtp.sendmail(sender, [recipient], msg.as_string())
  File "/Users/liamlynch/.pyenv/versions/3.7.13/lib/python3.7/smtplib.py", line 886, in sendmail
    raise SMTPRecipientsRefused(senderrs)
smtplib.SMTPRecipientsRefused: {'': (501, b'Error: Bad recipient address syntax')}
```

liamlynch-data replied:
Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/3f593c3e81ba6af3b561087b908a4222ca8ab47d)

MayeulRousselet commented on python-lib/dku_attachment_handling.py:
```diff
@@ -0,0 +1,46 @@
+        with attachment_ds.raw_formatted_data(format=request_fmt) as stream:
+            file_bytes = stream.read()
+        if attachment_type == "excel":
+            attachment_files.append(AttachmentFile()(attachment_ds.full_name + ".xlsx", "application",
```
Prefer using a class rather a NamedTuple

liamlynch-data replied:
done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/d94564d35ffc214777d231bc91872c8d123a4760

MayeulRousselet commented on python-lib/dku_attachment_handling.py:
```diff
@@ -0,0 +1,46 @@
+            attachment_files.append(AttachmentFile()(attachment_ds.full_name + ".xlsx", "application",
+                                                     "vnd.openxmlformats-officedocument.spreadsheetml.sheet", file_bytes))
+        else:
+            attachment_files.append(AttachmentFile()(attachment_ds.full_name + ".csv", "text", "csv", file_bytes))
```
Same comment as above

liamlynch-data replied:
done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/d94564d35ffc214777d231bc91872c8d123a4760

MayeulRousselet commented on python-lib/dku_email_client.py:
```diff
@@ -0,0 +1,112 @@
+        sender_to_use = None if self.channel.sender else sender
+        self.channel.send(self.project_id, [recipient], email_subject, email_body, attachments=files, plain_text=self.plain_text, sender=sender_to_use)
+
+    def quit(self):
```
Should be also declared in the abstract class

liamlynch-data replied:
Done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/ee8f64a5f31f04ab7a85974561e6dcee19aa45ad

MayeulRousselet commented on python-lib/dku_email_client.py:
```diff
@@ -0,0 +1,112 @@
+
+    def quit(self):
+        """ Do any disconnection needed"""
+        self.smtp.quit()
```
Should be also declared in the abstract class

liamlynch-data replied:
Done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/ee8f64a5f31f04ab7a85974561e6dcee19aa45ad

MayeulRousselet commented on python-lib/dku_email_client.py:
```diff
@@ -0,0 +1,112 @@
+        :param attachment_files:attachments as list of  AttachmentFile
+        """
+        # Any generic email body decoration goes here
+        email_body_to_send = email_body
+        if self.plain_text:
+            email_body_to_send = email_body + '\n\n'
```
This should be done in the `build_email`method or in the `send_mail_for_conrtact` method. 
Message client should only take care of sending already formatted emails, not format or alter the body.
Plus if you remove it, you don"t need a `send_email_impl` method, only an abstract `send_email` method

liamlynch-data replied:
Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/546ecdc32cd570ce50ffa5cb7109db18bb2e1858)

MayeulRousselet commented on python-lib/dku_email_client.py:
```diff
@@ -0,0 +1,112 @@
+    def __init__(self, plain_text, channel_id):
+        super().__init__(plain_text)
+
+        self.dss_client = dataiku.api_client()
```
You don"t need to store `dss_client`, you don"t use it after

liamlynch-data replied:
Done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/7aecefade387f1d3576cd34f8b721701044ccc6a

MayeulRousselet commented on python-lib/dku_email_client.py:
```diff
@@ -0,0 +1,112 @@
+
+    def __init__(self, plain_text, smtp_config):
+        super().__init__(plain_text)
+        self.smtp = smtplib.SMTP(smtp_config.smtp_host, port=smtp_config.smtp_port)
+        # Use TLS if set
+        if smtp_config.smtp_use_tls:
+            self.smtp.starttls()
+        # Use credentials if set
+        if smtp_config.smtp_use_auth:
+            self.smtp.login(str(smtp_config.smtp_user), str(smtp_config.smtp_pass))
```
all the init should be performed in a dedicated method, "performLogin" for instance

liamlynch-data replied:
Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/d66b102869e15faa70d737a5e1df68c31b08a8cb)

MayeulRousselet reviewed (changes requested):
Plugin works quite well, I have mainly concerns about maintenability and simplicity of the code. 
As a general comment, I found the logic around the use_boby_XXX very hard to follow. Maybe a simplification effort could be done here.

I've found the following issues:
1. If the attached dataset contains empty values, and you use `{{dataset.DATASET_NAME.html_table}}`, it will result in `NaN` value instead of empty values.
2. Very Edgy: If you share a dataset A from a project "External", and you have a dataset A in your project, you have 2 datasets A and External.A in the flow. Now, use these datasets as attachements, and refer to one of them in the body with `{{attachments.A.html_table}}`. DSS will pick one, with no possibility to choose. It should noted that `{{attachments.External.A.html_table}}` does not work.

liamlynch-data committed:
change version check method name and only pass format_params when necessary [sc-161134]

liamlynch-data committed:
pinned dependencies, jinja and xlsxWriter

liamlynch-data committed:
move logging line, convert contacts to dict earlier

liamlynch-data committed:
minor change of html condition

liamlynch-data committed:
change named tuples into classes

liamlynch-data committed:
change info logging to just log recipient

liamlynch-data committed:
changed with sender suffix so it is less confusing

liamlynch-data committed:
change direct smtp channel code to __DKU__DIRECT_SMTP__

liamlynch-data committed:
added quit() to AbstractMessageClient

liamlynch-data committed:
attachment buffer lines out and simplified messaging client code as per review suggestion

liamlynch-data committed:
removed unneeded api client member

liamlynch-data committed:
empty values in tables not NaN anymore

liamlynch-data commented:
> 1. If the attached dataset contains empty values, and you use `{{dataset.DATASET_NAME.html_table}}`, it will result in `NaN` value instead of empty values.

Addressed with
https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/1e3c70f126fcbdc95dbd0cae917398ccb9d65865

liamlynch-data committed:
handle foreign attachment datasets in templating, inc with the same name as local

liamlynch-data commented:
> 2. Very Edgy: If you share a dataset A from a project "External", and you have a dataset A in your project, you have 2 datasets A and External.A in the flow. Now, use these datasets as attachements, and refer to one of them in the body with `{{attachments.A.html_table}}`. DSS will pick one, with no possibility to choose. It should noted that `{{attachments.External.A.html_table}}` does not work.

Done in https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/2070fc6307fa6bdf4976ede41ebcc3fb1786dbac

For foreign datasets, you have to do:

```
{{ attachments.<project_key>.<ds_name>.html_table }}
```
E.g.

```
{{ attachments.DKU_TUTORIAL_BASICS_103.orders.html_table }}
```

I decided to using  the external project key here as I think it is simpler than choosing a special key and there could be more than one of the same name coming from different projects, so it is best to be unambiguous anyway (or even more edgy cases will break....)

liamlynch-data committed:
merge in origin/feature/dss12-sc-161126-plugin-add-jinja-templating-to-write-body

liamlynch-data committed:
Merge pull request #10 from dataiku/feature/dss12-sc-161134-plugin-add-a-settings-so-the-excel-attachments

adding option for excel conditional formating [sc-161134]

liamlynch-data committed:
make validation logic clearer as suggested in code rev

liamlynch-data committed:
move login of email clients to separate method

liamlynch-data committed:
split out send_email_for_contact method into others and removed use of globals

liamlynch-data committed:
removed residual send_email_for_contact method as was not really needed

liamlynch-data committed:
renamed email building methods to build_XXX

liamlynch-data commented:
Addressed all comments and retested including checking all supported versions of python 
Ready for you to take another look @MayeulRousselet (and @alexbourret you may want to look too)

MayeulRousselet reviewed (approved):
Ok for me!! 
Thanks for implementing the comments
